### ISSUE TITLE
What's the difference between the accuracy in the training script and one in the infer_all scirpt?

### ISSUE THREAD

**Issue by LaHauzel** [linminhtoo/neuralsym#3]
First thanks for your reimplementation! great work!

I would like to apply the code to my own dataset, but I found a huge gap between the accuracy in the training script and one in the infer_all script.

During the inference stage in the training script, I got the following output:
```  
test top-1 acc: 0.4992     
test top-2 acc: 0.6064, test top-3 acc: 0.6559,     
test top-5 acc: 0.7034, test top-10 acc: 0.7487,     
test top-20 acc: 0.7768, test top-50 acc: 0.8026,     
test top-100 acc: 0.8170     
```

However, when I used the `infer_all.sh`, the accuracy dropped rapidly:
```text
Test Top-1 accuracy:  6.358%
Test Top-3 accuracy:  7.160%
Test Top-5 accuracy:  7.253%
Test Top-10 accuracy:  7.275%
Test Top-20 accuracy:  7.278%
Test Top-50 accuracy:  7.280%
Test Top-100 accuracy:  7.280%
Test Top-200 accuracy:  7.280%
```

Which figure is more reliable? 
Any reply would be greatly appreciated!

**Comment by linminhtoo** [comment]
Hi, I'm happy to hear that my reimplementation has been of help.

In theory, the SMILES matching accuracy from infer_all.sh must be equal or higher than the template matching accuracy printed by the training script. This is because there is only one groundtruth template specified (which is not real since multiple templates can give the same, correct reactant SMILES). Put another way, neuralsym may have found an alternative template, which doesn't match the groundtruth template in the test set (thus lowering the template accuracy), but nevertheless gives the exact same reactant SMILES as the groundtruth template (thus increasing the SMILES accuracy.

In your case, I suspect something wrong is happening when the code is trying to string match the groundtruth reactant SMILES in your dataset against the reactant SMILES produced by applying the predicted template by neuralsym.

 Are the reactant SMILES in your test set canonicalised with the same rdkit version you're running neuralsym with? Are there any excess salts or other species in your groundtruth reactant SMILES?

**Comment by linminhtoo** [comment]
It would be helpful for debugging to print examples of reactions where the predicted template is correct but somehow the predicted reactant SMILES doesn't match what is notated in your groundtruth csv

**Comment by LaHauzel** [comment]
Great thanks for your detailed explanation and advice!
I have ensured that all the code is running with the same rdkit version.

Taking your advice, I tried to print some examples of reactions where the predicted template is correct but somehow the predicted reactant SMILES doesn't match. And it seemed that desired reactants could not be generated correctly with correct template.

Here is a example:
The reaction with atom mapping is: 
```text
[F:1][c:2]1[c:3]([N+:10](=[O:11])[O-:12])[cH:4][c:5]([F:9])[c:6]([F:8])[cH:7]1.[NH2:13][c:14]1[s:15][cH:16][cH:17][c:18]1[C:19]#[N:20]>>[c:2]1([NH:13][c:14]2[s:15][cH:16][cH:17][c:18]2[C:19]#[N:20])[c:3]([N+:10](=[O:11])[O-:12])[cH:4][c:5]([F:9])[c:6]([F:8])[cH:7]1
```
And the top-1 predicted reaction template, which is also the desired template, is:
```text
[#16&a:1]:[c:2]-[N&H1&D2&+0:3]-[c&H0&D3&+0:5](:[c:6]):[c:7]>>[#16&a:1]:[c:2]-[N&H2&D1&+0:3].[F&H0&D1&+0:4]-[c&H0&D3&+0:5](:[c:6]):[c:7]
```
These are their corresponding images:
![image](https://user-images.githubusercontent.com/25118553/184843273-3fbe6c50-2770-4369-be75-3af681579668.png)



I used the following function to get reactant:
```python
def gen_precs(templates_filtered, preds, phase_topk, task):
    i, prod_smi_nomap = task
    # generate predictions from templates
    precursors, dup_count = [], 0
    pred_temp_idxs = preds[i]
    for idx in pred_temp_idxs:
        template = templates_filtered[idx]

        try:
            rxn = rdchiralReaction(template)
            prod = rdchiralReactants(prod_smi_nomap)
            precs = rdchiralRun(rxn, prod)
            precursors.extend(precs)
        except Exception as e:
            print(e)
            continue

    # remove duplicate predictions
    seen = []
    for prec in precursors: # canonicalize all predictions
        prec = Chem.MolToSmiles(Chem.MolFromSmiles(prec), True)
        if prec not in seen:
            seen.append(prec)
        else:
            dup_count += 1

    if len(seen) < phase_topk:
        seen.extend(['9999'] * (phase_topk - len(seen)))
    else:
        seen = seen[:phase_topk]
    
    return precursors, seen, dup_count

```

Then an error occurred when running `rdchiralRun`: 
```text
Traceback (most recent call last):
  File "/nfs3-p1/LH/Synthesis/neuralsym/a.py", line 76, in <module>
    precs = rdchiralRun(rxn, prod)
  File "/home/LH/anaconda3/envs/Synthesis/lib/python3.6/site-packages/rdchiral/main.py", line 171, in rdchiralRun
    for m in outcome for a in m.GetAtoms() if a.HasProp('old_mapno')}
  File "/home/LH/anaconda3/envs/Synthesis/lib/python3.6/site-packages/rdchiral/main.py", line 171, in <dictcomp>
    for m in outcome for a in m.GetAtoms() if a.HasProp('old_mapno')}
KeyError: 4
```

It seems to be a rdchiral error?  Is there any method to solve it? 

Thanks again for your help!

**Comment by linminhtoo** [comment]
hi @LaHauzel, 
I do not remember having encountered myself before, unfortunately. does this happen for many of the reactions (and hence the main cause of the very low reactant SMILES-matching accuracy ?

I'm not terribly familiar with the inner workings of `rdchiral` so I would recommend you to create an Issue on `rdchiral`'s github repository for this one. 

https://github.com/connorcoley/rdchiral

**Comment by linminhtoo** [comment]
hi @LaHauzel,
I do not remember having encountered myself before, unfortunately. does this happen for many of the reactions (and hence the main cause of the very low reactant SMILES-matching accuracy ?

I'm not terribly familiar with the inner workings of rdchiral so I would recommend you to create an Issue on rdchiral's github repository for this one. Sorry I couldn't be of much help here.
https://github.com/connorcoley/rdchiral

Just to confirm, did you use the function in this repo to extract the template from the reaction SMILES? or did you extract them elsewhere?

**Comment by LaHauzel** [comment]
The reaction templates are all extracted by the provided script `prepare_data.py`.

I think I have solved the problem. With the reaction function provided by rdkit, I successfully generated the desired result. All of my modification is changing the following code
```python
rxn = rdchiralReaction(template)
prod = rdchiralReactants(prod_smi_nomap)
precs = rdchiralRun(rxn, prod)
precursors.extend(precs)
```
to
```python
rxn = AllChem.ReactionFromSmarts(template)
prod = [Chem.MolFromSmiles(prod_smi_nomap)]
precs = rxn.RunReactants(prod)
for prec in precs:
    prec = [Chem.MolToSmiles(p) for p in prec]
    precursors.append(".".join(prec))
```
This new function works as desired, resulting in a much better result which meets your view that the SMILES matching accuracy from infer_all.sh must be equal or higher than the template matching accuracy printed by the training script.
The final top-1 accuracy is ~53% and top-100 accuracy is ~85%.

Thanks again for your faithful reimplementation and help! You could close the issue now.

**Comment by linminhtoo** [comment]
hi, @LaHauzel , very happy hear that you found a fix. seems that the way I was using `rdchiral` was outdated. 53% top-1 accuracy is also very nice! all the best with your tinkering.

could I trouble you to create a PR to merge this fix into master? that would be very much appreciated :)

**Comment by LaHauzel** [comment]
> hi, @LaHauzel , very happy hear that you found a fix. seems that the way I was using `rdchiral` was outdated. 53% top-1 accuracy is also very nice! all the best with your tinkering.
> 
> could I trouble you to create a PR to merge this fix into master? that would be very much appreciated :)

I am glad you like it, but i am not sure if these two functions works exactly the same.
I tried the provided uspto-50k dataset with these two, and got different results, especially before removing duplicates:
```text 
(rdchiral version)
Avg # dups per product: 5.079888156580787

Calculating ranks before removing duplicates


Test Top-1 accuracy:  46.195%
Test Top-3 accuracy:  66.667%
Test Top-5 accuracy:  73.877%
Test Top-10 accuracy:  80.907%
Test Top-20 accuracy:  85.660%
Test Top-50 accuracy:  87.737%
Test Top-100 accuracy:  87.977%
Test Top-200 accuracy:  87.977%



Calculating ranks after removing duplicates


Test Top-1 accuracy:  46.195%
Test Top-3 accuracy:  67.525%
Test Top-5 accuracy:  74.536%
Test Top-10 accuracy:  82.085%
Test Top-20 accuracy:  86.279%
Test Top-50 accuracy:  87.897%
Test Top-100 accuracy:  87.977%
Test Top-200 accuracy:  87.977%
```
```text
(rdkit version)
Avg # dups per product: 5.506490912722189

Calculating ranks before removing duplicates


Test Top-1 accuracy:  29.778%
Test Top-3 accuracy:  43.659%
Test Top-5 accuracy:  48.672%
Test Top-10 accuracy:  53.685%
Test Top-20 accuracy:  56.980%
Test Top-50 accuracy:  58.618%
Test Top-100 accuracy:  58.818%
Test Top-200 accuracy:  58.818%



Calculating ranks after removing duplicates


Test Top-1 accuracy:  46.175%
Test Top-3 accuracy:  67.446%
Test Top-5 accuracy:  74.556%
Test Top-10 accuracy:  82.025%
Test Top-20 accuracy:  86.239%
Test Top-50 accuracy:  87.957%
Test Top-100 accuracy:  88.037%
Test Top-200 accuracy:  88.037%
```

If you could live with this difference, I'd happy to create a PR.

**Comment by linminhtoo** [comment]
that seems quite strange. did you have a chance to look at the duplicates with rdchiral vs rdkit? 

 i assume the duplicates here refer to duplicated reactant-sets predicted by `neuralsym`. the strangest thing that stands out to me immediately is that the top-1 accuracy increased after removing duplicates. if the duplicates are indeed true duplicates, then removing the top-N duplicated predictions (starting from the top-1 prediction) does not change the top-1 prediction. Therefore, the top-1 accuracy must be the same in both cases (which was true with the rdchiral version). This might point to a bug in either the de-duplication code, or the accuracy-calculation code.

**Comment by LaHauzel** [comment]
I see, it's the order between the reactants causing the problem. The output of the rdkit version may get multiple identical results that only differs in the order or the reactants, such as `c1ccccc1.O` and `O.c1ccccc1`. 
If the target SMILES was `c1ccccc1.O` and the generated reactants was `O.c1ccccc1`, it would fail to match by string comparison.
So I convert each output to canonical SMILES in advance, and the new reuslt is as follows, which is much normal:

```text
(rdkit version)
Avg # dups per product: 5.310964649490713

Calculating ranks before removing duplicates


Test Top-1 accuracy:  46.175%
Test Top-3 accuracy:  66.667%
Test Top-5 accuracy:  73.797%
Test Top-10 accuracy:  80.807%
Test Top-20 accuracy:  85.360%
Test Top-50 accuracy:  87.617%
Test Top-100 accuracy:  87.857%
Test Top-200 accuracy:  87.857%



Calculating ranks after removing duplicates


Test Top-1 accuracy:  46.175%
Test Top-3 accuracy:  67.486%
Test Top-5 accuracy:  74.496%
Test Top-10 accuracy:  81.945%
Test Top-20 accuracy:  86.079%
Test Top-50 accuracy:  87.777%
Test Top-100 accuracy:  87.857%
Test Top-200 accuracy:  87.857%
```

Here is the new version of `gen_precs`:

```python
def gen_precs(templates_filtered, preds, phase_topk, task):
    i, prod_smi_nomap = task
    # generate predictions from templates
    precursors, dup_count = [], 0
    pred_temp_idxs = preds[i]
    for idx in pred_temp_idxs:
        template = templates_filtered[idx]
        try:
            # rxn = rdchiralReaction(template)
            # prod = rdchiralReactants(prod_smi_nomap)
            # precs = rdchiralRun(rxn, prod)
            # precursors.extend(precs)
            rxn = AllChem.ReactionFromSmarts(template)
            prod = [Chem.MolFromSmiles(prod_smi_nomap)]
            precs = rxn.RunReactants(prod)
            precs = list(set([".".join([Chem.MolToSmiles(p) for p in prec] ) for prec in precs]))
            precs = list(Chem.CanonSmiles(prec) for prec in precs)
            precursors.extend(precs)
        except:
            continue

    # remove duplicate predictions
    seen = []
    for prec in precursors: # canonicalize all predictions
        try:
            prec = Chem.MolToSmiles(Chem.MolFromSmiles(prec), True)
        except:
            prec = None
        if prec and prec not in seen:
            seen.append(prec)
        else:
            dup_count += 1

    if len(seen) < phase_topk:
        seen.extend(['9999'] * (phase_topk - len(seen)))
    else:
        seen = seen[:phase_topk]
    
    return precursors, seen, dup_count

```

**Comment by linminhtoo** [comment]
Cool! Glad that you found and fixed the bug. Would you be willing to make a PR for this?

**Comment by LaHauzel** [comment]
> Cool! Glad that you found and fixed the bug. Would you be willing to make a PR for this?

Of course, I have created a PR and you can check it now.

**Comment by linminhtoo** [comment]
addressed by #4
